Fix wiki editor so that hyperlink function adds link to highlighted text

kch8qx opened the pull request:
[OSF-5528]

whit537 commented on website/addons/wiki/static/pagedown-ace/Markdown.Editor.js:
```diff
@@ -1837,14 +1837,15 @@ var Range = ace.require('ace/range').Range;
 
     commandProto.addLinkDef = function (chunk, linkDef) {
-
+        
```
The whitespace changes are a little distracting. Good practice is to configure one's text editor to trim trailing whitespace for most file types.

whit537 replied:
(That'll introduce noise in the opposite direction for this particular PR, since the file in question already includes a number of lines with trailing whitespace. Such is hygiene?)

kch8qx replied:
whoops, I can fix the whitespace and comments that were accidentally left in and update the PR now

whit537 replied:
Cool, thanks. :)

kch8qx commented:
@whit537 so I'm not sure why it looks like I deleted and then added whitespace in the PR, since that whitespace already existed before? I did remove any whitespace that I added (that doesn't already exist on this file on develop) and the few comments that I accidentally left in for the last PR

whit537 commented:
:monkey_face:

whit537 commented:
:banana: :coffee: :chocolate_bar: :speak_no_evil:

whit537 commented:
Okay, sorry, having too much fun. :)

whit537 commented:
It's easy enough to suppress the whitespace changes in the diff on GitHub by adding [`?w=1`](https://github.com/blog/967-github-secrets) to the URL, like so:

https://github.com/CenterForOpenScience/osf.io/pull/4909/files?w=1

That said, it'd be nice to trim trailing whitespace in this file to avoid having the whitespace there in the first place.

whit537 commented:
What text editor are you using? Can you look into configuring it to trim trailing whitespace for certain filetypes? That's a good habit to get into.

whit537 commented:
It kinda gives me the willies that we have an inline fork of [pagedown-ace](https://github.com/benweet/pagedown-ace). :flushed:

whit537 commented:
> trim trailing whitespace

(We also don't want the newline changes.)

whit537 commented:
I'm comparing their current [`addLinkDef`](https://github.com/benweet/pagedown-ace/blob/c9d41937e1e3b4e46f823856a47f9eb22a8108e1/Markdown.Editor.js#L1811-L1866) with [ours](https://github.com/CenterForOpenScience/osf.io/blob/75d9bfbf23bb10cea6137751879a36d25a4f36f2/website/addons/wiki/static/pagedown-ace/Markdown.Editor.js#L1839-L1901) ... maybe they already fixed this bug? (I'm assuming that I've properly identified the source of our fork.)

whit537 commented:
Looks like not (`diff -rub`):

```diff
--- theirs.txt	2016-02-25 15:56:45.000000000 -0500
+++ ours.txt	2016-02-25 15:57:01.000000000 -0500
@@ -44,12 +44,19 @@
         chunk.after = chunk.after.replace(regex, getLink);
 
         if (chunk.after) {
+            if (chunk.selection) {
+                chunk.selection = '[' + chunk.selection + '][' + refOut + ']';
+            }
             chunk.after = chunk.after.replace(/\n*$/, "");
         }
-        if (!chunk.after) {
+        else if (!chunk.after) {
+            if (chunk.selection) {
+                chunk.selection = '[' + chunk.selection + '][' + refOut + ']';
+            }
             chunk.selection = chunk.selection.replace(/\n*$/, "");
         }
 
+
         chunk.after += "\n\n" + defs;
 
         return refOut;
```

whit537 commented:
Issues are turned off on the `pagedown-ace` repo. :hear_no_evil: : For the upstream project we are directed to Google Code, which is in sunset. :see_no_evil: I've downloaded the ZIP that Google still makes available. Here are the changes that were (presumably) made in the port to Ace:

```diff
--- upstream.txt	2016-02-25 16:02:20.000000000 -0500
+++ theirs.txt	2016-02-25 16:02:33.000000000 -0500
@@ -10,55 +10,6 @@
         var defs = "";
         var regex = /(\[)((?:\[[^\]]*\]|[^\[\]])*)(\][ ]?(?:\n[ ]*)?\[)(\d+)(\])/g;
         
-        // The above regex, used to update [foo][13] references after renumbering,
-        // is much too liberal; it can catch things that are not actually parsed
-        // as references (notably: code). It's impossible to know which matches are
-        // real references without performing a markdown conversion, so that's what
-        // we do. All matches are replaced with a unique reference number, which is
-        // given a unique link. The uniquifier in both cases is the character offset
-        // of the match inside the source string. The modified version is then sent
-        // through the Markdown renderer. Because link reference are stripped during
-        // rendering, the unique link is present in the rendered version if and only
-        // if the match at its offset was in fact rendered as a link or image.
-        var complete = chunk.before + chunk.selection + chunk.after;
-        var rendered = this.converter.makeHtml(complete);
-        var testlink = "http://this-is-a-real-link.biz/";
-        
-        // If our fake link appears in the rendered version *before* we have added it,
-        // this probably means you're a Meta Stack Exchange user who is deliberately
-        // trying to break this feature. You can still break this workaround if you
-        // attach a plugin to the converter that sometimes (!) inserts this link. In
-        // that case, consider yourself unsupported.
-        while (rendered.indexOf(testlink) != -1)
-            testlink += "nicetry/";
-        
-        var fakedefs = "\n\n";
-
-        // the regex is tested on the (up to) three chunks separately, and on substrings,
-        // so in order to have the correct offsets to check against okayToModify(), we
-        // have to keep track of how many characters are in the original source before
-        // the substring that we're looking at. Note that doLinkOrImage aligns the selection
-        // on potential brackets, so there should be no major breakage from the chunk
-        // separation.
-        var skippedChars = 0;
-
-        var uniquified = complete.replace(regex, function uniquify(wholeMatch, before, inner, afterInner, id, end, offset) {
-            skippedChars += offset;
-            fakedefs += " [" + skippedChars + "]: " + testlink + skippedChars + "/unicorn\n";
-            skippedChars += before.length;
-            inner = inner.replace(regex, uniquify);
-            skippedChars -= before.length;
-            var result = before + inner + afterInner + skippedChars + end;
-            skippedChars -= offset;
-            return result;
-        });
-        
-        rendered = this.converter.makeHtml(uniquified + fakedefs);
-        
-        var okayToModify = function(offset) {
-            return rendered.indexOf(testlink + offset + "/unicorn") !== -1;
-        }
-        
         var addDefNumber = function (def) {
             refNumber++;
             def = def.replace(/^[ ]{0,3}\[(\d+)\]:/, "  [" + refNumber + "]:");
@@ -70,12 +21,8 @@
         //    of regex, inner is always a proper substring of wholeMatch, and
         // b) more than one level of nesting is neither supported by the regex
         //    nor making a lot of sense (the only use case for nesting is a linked image)
-        var getLink = function (wholeMatch, before, inner, afterInner, id, end, offset) {
-            if (!okayToModify(skippedChars + offset))
-                return wholeMatch;
-            skippedChars += offset + before.length;
+        var getLink = function (wholeMatch, before, inner, afterInner, id, end) {
             inner = inner.replace(regex, getLink);
-            skippedChars -= offset + before.length;
             if (defsToAdd[id]) {
                 addDefNumber(defsToAdd[id]);
                 return before + inner + afterInner + refNumber + end;
@@ -83,18 +30,14 @@
             return wholeMatch;
         };
 
-        var len = chunk.before.length;
         chunk.before = chunk.before.replace(regex, getLink);
-        skippedChars += len;
         
-        len = chunk.selection.length;
         if (linkDef) {
             addDefNumber(linkDef);
         }
         else {
             chunk.selection = chunk.selection.replace(regex, getLink);
         }
-        skippedChars += len;        
 
         var refOut = refNumber;
 
@@ -111,4 +54,3 @@
 
         return refOut;
     };
-
```

whit537 commented:
Does the `skippedChars` removal explain why this was broken in the ACE port?

whit537 commented:
[OSF-5528](https://openscience.atlassian.net/browse/OSF-5528)

whit537 commented:
Confirmed broken on `develop`:

![screen shot 2016-02-25 at 4 13 05 pm](https://cloud.githubusercontent.com/assets/134455/13334656/2738af5c-dbdb-11e5-91b8-c2d25f28d35c.png)

whit537 commented:
Hmmm ... what I get on this branch is not what I would expect either, however.

![screen shot 2016-02-25 at 4 16 12 pm](https://cloud.githubusercontent.com/assets/134455/13334667/32e4c9f8-dbdb-11e5-8217-fc890bf77eaf.png)

whit537 commented:
I expected an HTML preview under "View". Is that wrong?

whit537 commented:
At any rate, `doo` shouldn't be in the link. What exactly is the expected output here?

whit537 commented:
##### Punchlist (for @kch8qx)

 - [ ] remove whitespace changes—https://github.com/CenterForOpenScience/osf.io/pull/4909#issuecomment-188975836
 - [ ] confirm that we're not expecting HTML under "View"—https://github.com/CenterForOpenScience/osf.io/pull/4909#issuecomment-188993041
 - [ ] fix bug in multi-word link text—https://github.com/CenterForOpenScience/osf.io/pull/4909#issuecomment-188993215

whit537 commented:
Alright @kch8qx, I've started [punchlist](https://github.com/CenterForOpenScience/osf.io/pull/4909#issuecomment-188993707) for you. :basketball: is in your court. Lemme know when you've addressed those issues and I'll take a look! :-)

whit537 commented:
IIIIIIIIIIIII MISTOOK YOU FOR SOMEONE ELSE I AM VERY EMBARRASSED AND I AM SORRY. :flushed: :see_no_evil:

kch8qx commented:
@whit537 
Working on fixing the whitespace changes now, but as far as the "Floober doo" example:

* Confirmed that we wouldn't expect to see an HTML preview under 'View' in the screenshot example. Even on production, this doesn't show an HTML preview, because the URL hasn't been typed correctly (includes spaces). I maintained that behavior when adding links by highlighting text.
* As far as "fix bug in multi-word link text" I would assume that we wouldn't want to allow the user to form links like this, and wouldn't want to just chop off anything after whitespace and make the rest a link. If we did that, a user could accidentally put a space into their URL, see that the link "works" (but goes to the truncated URL before the whitespace) and not notice the error. If we leave it as is, they could see the link didn't work under 'View', notice their accidental whitespace, remove it, and then see the link works correctly.
* Last item on checklist, I'm confused by what you mean "determine whether original code had this bug"? On production, links that are formatted with [1]: http://osf.io /doo don't work, is that what you meant? Confused by which bug you are referring to with "this bug"

whit537 commented:
Multiple words in link text confirmed working now. Not sure what I did before to get the result at https://github.com/CenterForOpenScience/osf.io/pull/4909#issuecomment-188992980.

![screen shot 2016-03-02 at 3 46 28 pm](https://cloud.githubusercontent.com/assets/134455/13474913/197e4d60-e08e-11e5-9e81-970ee71be40e.png)

whit537 commented:
I'm still seeing extraneous whitespace changes, but other than that I'm satisfied here.

:+1:

sloria commented:
Trimmed trailing whitespace when I merged.

whit537 commented:
!m @sloria
